src/poetry/repositories/repository_pool.py

```diff
@@ -0,0 +1,136 @@
+
+    def _contains_priority(self, priority: RepositoryPriority) -> bool:
+        return any(
+            repo_prio == priority for _, repo_prio in self._repositories.values()
+        )
+
+    def has_repository(self, name: str) -> bool:
+        return name.lower() in self._repositories
+
+    def repository(self, name: str) -> Repository:
+        name = name.lower()
+        if name in self._repositories:
+            return self._repositories[name][0]
+        raise ValueError(f'Repository "{name}" does not exist.')
+
+    def add_repository(
+        self, repository: Repository, default: bool = False, secondary: bool = False
+    ) -> RepositoryPool:
+        """
+        Adds a repository to the pool.
+        """
+        repository_name = repository.name.lower()
+        if repository_name in self._repositories:
+            raise ValueError(
+                "A repository with (case-independent) name"
+                f" {repository_name} was already added"
+            )
+
+        if default and self.has_default():
+            raise ValueError("Only one repository can be the default")
+
+        if default:
+            repository_priority = RepositoryPriority.DEFAULT
+        elif secondary:
+            repository_priority = RepositoryPriority.SECONDARY
+        else:
+            repository_priority = RepositoryPriority.PRIMARY
+        self._repositories[repository_name] = (repository, repository_priority)
+        return self
+
+    def remove_repository(self, repository_name: str) -> RepositoryPool:
+        del self._repositories[repository_name.lower()]
+        return self
+
+    def package(  # type: ignore[override]
+        self,
+        name: str,
+        version: Version,
+        extras: list[str] | None = None,
+        repository_name: str | None = None,
+    ) -> Package:
+        if repository_name and not self._ignore_repository_names:
+            return self.repository(repository_name).package(
+                name, version, extras=extras
+            )
+
+        for repo in self.repositories:
+            try:
+                return repo.package(name, version, extras=extras)
+            except PackageNotFound:
+                continue
+        raise PackageNotFound(f"Package {name} ({version}) not found.")
+
+    def find_packages(self, dependency: Dependency) -> list[Package]:
+        repository_name = dependency.source_name
+        if repository_name and not self._ignore_repository_names:
+            return self.repository(repository_name).find_packages(dependency)
+
+        packages: list[Package] = []
+        for repo in self.repositories:
+            packages += repo.find_packages(dependency)
+        return packages
+
+    def search(self, query: str) -> list[Package]:
+        from poetry.repositories.legacy_repository import LegacyRepository
+
+        results: list[Package] = []
+        for repository in self.repositories:
+            if isinstance(repository, LegacyRepository):
```

Comment:
Instead of special casing here, it's likely better/cleaner to just return an empty list in `LegacyRepository`.

Reply:
Fixed. We might even clean up `LegacyRepository` further by removing all references to `self._packages` -- they seem to never get populated.